### DISCUSSION TITLE
Request acknowledgement (especially Kademlia)

### DISCUSSION THREAD

**Discussion by nazar-pc** [libp2p/rust-libp2p#3411] [General]
We have a challenge with provider announcements in Kademlia. When requests come in, we basically have to process them right away, there doesn't seem to be any way to tell library that request needs more time to be processed.

Specifically, we need to manage enormous number of announcements and we use disk-based storage, which means no guarantee on how much time it takes to write to disk, so we need to do that in a separate thread. Moreover, we do other time consuming things in addition to that. What this means is that when we get a storm of announcements there is no way for us to tell announcer to wait a little while we're processing previous announcements.

As the result we have a bounded queue that fills up quickly and after that we just ignore majority of announcements simply to avoid running out of memory, which is not ideal because we can't discover the content later due to announcements not being stored.

https://github.com/libp2p/rust-libp2p/issues/3035 is kind of sort of related. While it talks about errors, it was already mentioned by @mxinden that we would likely also want to make those async. I'm wondering if the issue described here is covered by the same task or more tickets need to be created.

This is the next major issue for Subspace and I'm willing to dedicate time implementing/fixing it given some guidance on how to approach it.

**Comment by thomaseizinger** [comment]
Effectively, what you are asking for is back-pressure, ideally on newly opened streams all the way to the other the peer. See https://github.com/libp2p/rust-libp2p/issues/3078.

There are a lot of moving parts here but I'll try to summarize:

- Our `StreamMuxer` interface can in theory already do back-pressure: The `poll_inbound` interface allows us to not accept new inbound streams if we cannot handle them at the moment.
- Kademlia states in its spec that streams can be reused. This should probably go away such that we can piggy back on the back-pressure of new inbound streams: Not being able to handle any more inbound streams will automatically mean that we cannot handle any more inbound kademlia requests.
- The `ConnectionHandler` interface currently has **no ability** to signal whether it can handle more streams. It is simply given streams as they come available. This is one of the holes in our back-pressure chain.
- One issue in solving the above is that we need to accept a stream from the remote and run `multistream-select` on it before we even know which `ConnectionHandler` to ask whether it can handle more streams.

**Reply by thomaseizinger** [reply]
> * Kademlia states in its spec that streams can be reused. This should probably go away such that we can piggy back on the back-pressure of new inbound streams: Not being able to handle any more inbound streams will automatically mean that we cannot handle any more inbound kademlia requests.

A Kademlia-specific solution would be to actually embrace this feature. At the moment, `rust-libp2p` does not reuse streams although the receiving end is written to handle it. We might be able to have some wins here by actually reusing streams. Once they are established, we can provide back-pressure by only polling for new inbound messages when we are ready to handle them which will effectively only be once you have processed a previous one.

**Reply by thomaseizinger** [reply]
See https://github.com/libp2p/rust-libp2p/blob/0c94237e16707380857cd04b9bd76b6d2979e7a5/protocols/kad/src/handler.rs#L913 and https://github.com/libp2p/rust-libp2p/blob/0c94237e16707380857cd04b9bd76b6d2979e7a5/protocols/kad/src/handler.rs#L936.

In both of these cases, we close the substream after having sent a message. We'd have to introduce a new state in [`OutboundSubstreamState`](https://github.com/libp2p/rust-libp2p/blob/0c94237e16707380857cd04b9bd76b6d2979e7a5/protocols/kad/src/handler.rs#L150-L169) called `Idle` that only holds the stream. Every time we want to send a message, we would then have to check if we have an existing stream in this state and reuse that one instead of opening a new one.

This doesn't fully give us back-pressure yet though because a `NetworkBehaviour` can still flood a `ConnectionHandler` with messages.

Again, we might be able to for the moment build something specifically for `Kademlia` where we keep track of how many requests we have sent to a particular connection. Once we have that information in the `Kademlia` behaviour, we can back-pressure to the user and refuse to take more requests in case all connections are busy.

**Reply by thomaseizinger** [reply]
> This doesn't fully give us back-pressure yet though because a NetworkBehaviour can still flood a ConnectionHandler with messages.
> 
> Again, we might be able to for the moment build something specifically for Kademlia where we keep track of how many requests we have sent to a particular connection. Once we have that information in the Kademlia behaviour, we can back-pressure to the user and refuse to take more requests in case all connections are busy.

An alternative to this might be to have a buffer (with configurable size) in the `ConnectionHandler` where we keep incoming messages and thus more easily handle these kind of bursts.

**Reply by nazar-pc** [reply]
I was hoping there is a simpler mid-term solution comparing to full backpressure, but I'm open to do progress towards backpressure as that is something that has to happen eventually regardless.

On Kademlia side we feel pretty good without stream reuse after https://github.com/libp2p/rust-libp2p/pull/3287, it doesn't seem to be the highest priority even though it should improve things a bit.

If we implement stream reuse in Kademlia though, does it mean we'll be able to make Kademlia wait for acknowledgement from an app before reading the next message?

It felt like Kademlia on sending side shouldn't just fire and forget and if it indeed the case, we should be able to have some kind of handle in the request event that unless dropped prevents requests from being acknowledged, so I can move it into processing pipeline and only drop when ready to do so.

**Reply by thomaseizinger** [reply]
Yeah the idea would be to store all to-be-sent requests in a buffer in the `ConnectionHandler` and only send them as streams become available. That would mean a max of 32 at any one time.

Said buffer would have to have an upper bound to avoid unlimited memory growth but this could be configurable and as large as 1000 elements to handle such kind of bursts.

One issue we might run into is that `ADD_PROVIDER` currently does not have a response, meaning we cannot delay the sender to not immediately send another `ADD_PROVIDER`.

I wonder if we can add one in a backwards-compatible way.

**Reply by nazar-pc** [reply]
Yes, the fact that it doesn't have a response is something I was wondering about actually. So looks like it will send messages until it fills some buffer in the muxer or will it break the connection? I don't think I'm fully aware of the implications. Because on receiver side we could stop reading byte until we're done processing what we have already.

**Reply by thomaseizinger** [reply]
`yamux` has support for back-pressure on a stream level so yes, we won't be able to write any more messages if the remote hasn't read from the stream: https://github.com/libp2p/rust-yamux/blob/e6e10f82628847f0a789a63019328dfcc0324af4/src/connection/stream.rs#L345-L349

**Reply by nazar-pc** [reply]
Thanks, I think at least that would be nice to have in Kademlia: make it not read new messages until old ones are processed fully. This should be backwards compatible change.

**Reply by thomaseizinger** [reply]
However, I don't think it is going to make a difference in your situation. Your issue seems to be that you can't keep up with processing the outputs of the `Kadmelia` behaviour. The question is, why are you polling the swarm if you can't handle another event?

If your queue for processing events is full, stop polling the `Swarm`. That should fill up the queue between `ConnectionHandler` and `NetworkBehaviour` which means we will stop polling the connection, which means we will stop reading messages from the stream.

**Reply by nazar-pc** [reply]
Well, there are different things that use Swarm. Since it is one loop for all connections, we can't just stop polling it indefinitely as it'll break other assumptions of the protocol. We need to be able to throttle specific peers, ideally specific streams from them rather than pause all networking communication. Especially because part of the processing is actually caching things and re-announcing it outselves, with paused networking it'll not happen :upside_down_face:

**Reply by thomaseizinger** [reply]
I just looked again at the implementation. Are you using `KademliaStoreInserts::Unfiltered`?

**Reply by nazar-pc** [reply]
No, we use `KademliaStoreInserts::FilterBoth` and handle things explicitly

**Reply by thomaseizinger** [reply]
With `KademliaStoreInserts::Unfiltered`, the `NetworkBehaviour` will block until the request is processed. I am not sure we can do more than that?

> We need to be able to throttle specific peers, ideally specific streams from them rather than pause all networking communication.

The problem in making this happen is that we'd need to co-locate the processing of a message with the connection. Currently, the events from the connection are passed up to the behaviour which is in charge of all connections.

If you are up for some experiments, I can suggest the following:

- Fork `libp2p-kad`
- Store the `RecordStore` within `ConnectionHandler` as an `Arc<Mutex<RecordStore>>`
- Process the incoming messages directly in the `ConnectionHandler`
- Probably make the `RecordStore` interface `async`

With this design, a connection that is getting hammered with requests will backpressure to the remote because the entire task is suspended in reading / writing to the `RecordStore`.

You can also go for a slight variation where the `RecordStore` does not take `&mut self` and instead uses interior mutability (or channels).

**Reply by thomaseizinger** [reply]
Another nice thing about the above design is that you can express the entire protocol as `async fn`s. Pseudo-code:

```
async fn handle_inbound_stream(stream, record_store) {
	let msg = stream.next();

	match msg {
		GetRecord => {
			let record = record_store.get();
			stream.send(record).await;
		}
		...
	}
}
```

In the current implementation, there is a lot of message passing between the `NetworkBehaviour` and the `ConnectionHandler` because the `RecordStore` is only accessible from the former.

**Reply by thomaseizinger** [reply]
The message passing way of solving this would be to introduce a message from the `NetworkBehaviour` to the `ConnectionHandler` once a provider is added. The `ConnectionHandler` could then wait with reading further messages from a stream until it has received this confirmation.

**Reply by nazar-pc** [reply]
I'd really prefer to not maintain a fork, especially long-term, and rather improve upstream. But I may go that route in the meantime since it sound reasonably straightforward and not too bad to maintain.

> The message passing way of solving this would be to introduce a message from the NetworkBehaviour to the ConnectionHandler once a provider is added. The ConnectionHandler could then wait with reading further messages from a stream until it has received this confirmation.

This feels like a more idiomatic solution in rust-libp2p.

**Reply by thomaseizinger** [reply]
> I'd really prefer to not maintain a fork, especially long-term, and rather improve upstream. But I may go that route in the meantime since it sound reasonably straightforward and not too bad to maintain.

I can't promise that we'd accept it upstream but I also wouldn't rule it out, esp. if it turns out significantly simpler and allows to easily backpressure individual streams / connections.

> > The message passing way of solving this would be to introduce a message from the NetworkBehaviour to the ConnectionHandler once a provider is added. The ConnectionHandler could then wait with reading further messages from a stream until it has received this confirmation.
> 
> This feels like a more idiomatic solution in rust-libp2p.

It has definitely historically been the way things get implemented in here (with reason). That doesn't mean we shouldn't experiment with other designs.

**Reply by nazar-pc** [reply]
Okay, I'll experiment with it and let you know if I have something meaningful to share.
Thanks for suggestions!

**Reply by thomaseizinger** [reply]
Another idea just came to my mind:

- Use `Arc<Mutex<RecordStore>>` in `ConnectionHandler`
- `RecordStore` interface is _not_ async but has `poll` functions:
	- `RecordStore::poll_ready_add_provider`
	- `RecordStore::add_provider`

This would allow us to use a regular `std::sync::Mutex` because the lock times would be extremely short:

1. Lock Mutex
2. Call poll
3. If ready, add provider
4. If pending, return pending (i.e. suspend `ConnectionHandler`)

Predictable lock times and not holding locks across `await` points make it much less likely that we run into deadlocks which makes the use of a `Mutex` less bad.

**Comment by mxinden** [comment]
Sorry for only commenting here now. Difficult topic. Thanks for the elaborate discussion. A couple of thoughts. I will discuss this more synchronously with @thomaseizinger tomorrow.

- I don't think reusing streams is the way to go. Kademlia requests are small. They will most likely not saturate a stream receive window, thus no backpressure. Reusing streams implies head-of-line blocking.
- Async or poll style `RecordStore` sounds reasonable.
- `Arc<Mutex<RecordStore>>` I would prefer `RecordStore` only, having the implementer of `RecordStore` decide whether they use a `Mutex` or something else under the hood.
- We should rely on stream-creation backpressure of the underlying Muxer implementation. For `libp2p-quic` that should give us a [concurrency limit of `100`](https://docs.rs/quinn-proto/0.9.2/src/quinn_proto/config.rs.html#258), i.e. at most 100 Kademlia requests in-flight at any given point in time. As long as `libp2p-kad` does not drop an inbound stream (e.g. waiting for `RecordStore`) `libp2p-quic` does not allow the remote node a new stream and thus the remote producer (`libp2p-kad`) is backpressured.
- While ideally we would have a backpressure mechanism for `ConnectionHandler` to deny new streams, this is tricky. As mentioned above, at first we don't know what protocol a new incoming stream is for, as `multistream-select` hasn't yet run.
